Writeinpublic button logic

deanmalan opened the pull request:
[Pivotal](https://www.pivotaltracker.com/story/show/172920709)

[Related PR in WriteInPublic](https://github.com/OpenUpSA/writeinpublic/pull/10)

This PR aims to:
- Ensure only **current members of the National Assembly** for whom we have **email addresses** can be contacted via WriteInPublicin PA on the [Write to MP page](https://www.pa.org.za/write/recipients/), [Profile page](https://www.pa.org.za/person/kevin-john-mileham/) and the [Popolo JSON](pa.org.za/api/national-assembly/popolo.json) route that WriteInPublic consumes.
- On the [draft page](https://www.pa.org.za/write/draft/) ensure that the person's email address has been saved to the WriteInPublic database by making a request to the WriteInPublic API.

On the draft page, when the person can't be found in WriteInPublic or the WriteInPublic database has no email address for the person, the following message is displayed:

![2020-06-24-12:31](https://user-images.githubusercontent.com/4767109/85540498-5aacd280-b617-11ea-838a-be3b767d203d.png)

deanmalan committed:
Show write to MP button if and only if a person is a member of parliament

deanmalan committed:
Add preferred_email_address property to Person model

deanmalan committed:
Rename property to preferred_email

deanmalan committed:
Check whether person is reachable in WIP on draft form

deanmalan committed:
Use preferred_email in json file

deanmalan committed:
Revert "Use preferred_email in json file"

This reverts commit 88436293b02c5cda919bc8763b125eaebfe3fbec.

deanmalan committed:
Add a for_write_to_mp method to PersonQuerySet

deanmalan committed:
Check for email address for Write to MP button

deanmalan committed:
Don't check for everypolitician_uuid before showing email

deanmalan committed:
Remove extra logging

deanmalan committed:
Remove extra lines and print statement

deanmalan committed:
Remove WIP configuration slug from person view

deanmalan committed:
Change message a bit

deanmalan committed:
Remove non-contactable recipients

deanmalan committed:
Change message a bit

deanmalan committed:
Add tests for Write in Public contactable

deanmalan committed:
Check if has ever been member of NA for messages tab

deanmalan committed:
Use has_contacts instead of contactable

deanmalan committed:
Filter by instance id and fix test for has_contacts

jbothma commented on pombola/south_africa/templates/core/person_detail.html:
```diff
@@ -316,7 +316,7 @@ <h3>Party</h3>
       </ul>
 
-    {% if person.everypolitician_uuid and email_contacts %}
+    {% if email_contacts %}
```
I think we should replace everypolitician_uuid with has_ever_been_member_of_national_assembly

jbothma commented on pombola/writeinpublic/views.py:
```diff
@@ -218,11 +212,32 @@ def get_template_names(self):
+                        else:
+                            contactable.append(person)
+                    except Exception:
+                        not_contactable.append(person)
```
always log exceptions, ideally in a way sentry will find them. Looks like logger.exception will do the right thing
https://docs.sentry.io/platforms/python/logging/
This is for things like http exceptions, right? we want the user to proceed but we want to know the system is funky

jbothma commented on pombola/core/models.py:
```diff
@@ -337,6 +337,13 @@ def is_politician(self, when=None):
         # FIXME - Don't like the look of this, rather a big subquery.
         return self.filter(position__in=Position.objects.all().current_politician_positions(when))
 
+    def for_write_to_mp(self):
```
```suggestion
    def current_mps_with_email(self):
```

jbothma commented on pombola/writeinpublic/client.py:
```diff
@@ -35,6 +35,28 @@ def __init__(self, params, person):
+            setattr(self, key, params.get(key))
+
+        self._params = params
+        self.adapter = adapter
```
I'm not totally sure what's happening here - are we generating a duck-typed pombola core person from a a recipient? or am I being dim?

what purpose does this class serve?

jbothma replied:
I just have the suspicion we're mixing in a concept here that doesn't belong in the writeinpublic client and might not need to exist

jbothma replied:
I'm so confused, sorry :) what's an adapter? I just don't understand the relevance of this code to the change and you've probably figured this all out

jbothma commented on pombola/writeinpublic/client.py:
```diff
@@ -101,3 +123,33 @@ def get_messages(self, person_popolo_uri):
+            'has_contacts': 'true',
+            'instance_id': self.instance_id
+        }
+        write_to_mp_people = self.get_person(person, filters)
```
```suggestion
        persons_with_contacts = self.get_person(person, filters)
```

jbothma commented on pombola/writeinpublic/tests.py:
```diff
@@ -126,6 +180,43 @@ def test_get_messages(self, m):
+            'identifiers__scheme': ['popolo:person'],
+            'identifiers__identifier': [str(person.id)],
+            'has_contacts': ['true'],
+            'instance_id': ['42']
```
in theory you can check multiple contacts at once to reduce the number of queries because you can pass a param multiple times, but that's a premature optimisation

jbothma commented on pombola/writeinpublic/views.py:
```diff
@@ -218,11 +212,32 @@ def get_template_names(self):
         return [self.adapter.get_templates()[self.steps.current]]
 
     def get_context_data(self, form, **kwargs):
+        step = kwargs.get('step')
         context = super(WriteInPublicNewMessage, self).get_context_data(form=form, **kwargs)
         context['message'] = self.get_cleaned_data_for_step('draft')
         recipients = self.get_cleaned_data_for_step('recipients')
+
         if recipients is not None:
-            context['persons'] = recipients.get('persons')
+            persons = recipients.get('persons')
+            if step == 'draft':
+                # On the draft step, check if all recipients are contactable
+                # in WriteInPublic
+                not_contactable = []
+                contactable = []
+                for person in persons:
+                    try:
+                        if not self.client.get_person_is_contactable(person):
+                            not_contactable.append(person)
+                        else:
```
is there already a test for this? I'm happy just to have one success case test, the rest can be handled with regression tests as errors come in

deanmalan replied:
Yup, there is this test: https://github.com/OpenUpSA/pombola/pull/47/files#diff-9aa30a36f8728d1db5955b769cc6ac3eR224

deanmalan committed:
Log WriteInPublic API call exceptions

deanmalan committed:
Check if person has been member of NA before displaying email

deanmalan committed:
Don't display the email address button for inactive MPs

In an email from Rashaad he said that for inactive MPs we can keep the email
address, but it must not the emphasized on the profile in anyway.

deanmalan committed:
Replace for_write_to_mp with current_mps_with_email

deanmalan committed:
Replace write_to_mp_people with persons_with_contacts

deanmalan committed:
Test commit

deanmalan committed:
Revert "Test commit"

This reverts commit 1b4ddb8dccf48d758f1a57d15952efa306aa4e41.

deanmalan committed:
Remove Person class from WIP client

deanmalan committed:
Run writeinpublic tests on Travis

deanmalan committed:
Fix Writeinpublic get_person test

deanmalan committed:
Test email address and Write to MP button on profile

deanmalan committed:
Test the a tag specifically

jbothma reviewed (approved):
Awesome. Thanks!
